### PR TITLE
Use resource manager

### LINKED ISSUE #164: Implement ResourceManager based on Components
With the newly introduced components, the requirements for the resource manager are as follows:

* Remodel Hardware Components
  * Remove existing "joint" and "sensor" components
  * Move hardware resources into components namespace and use as such
  * Update component parser to reflect changes
* Remodel Component Interfaces
  * Introduce handles for command and state interfaces to hardware components
  * Update hardware classes to instantiate these handles given a list of interface strings
* Allocate initial memory for each of these components - these components are fix on runtime
* Provide Resource Loaning mechanism:
  * Handles to components can be uniquely loaned/assigned to controllers
  * Components are automatically released when controller is shutdown/unloaded
* Controller Manager iterates over all components for read/write
* Change Controller Interface to access "SingleValueComponent" through resource manager

Changes will be made according to the design draft here: https://github.com/ros-controls/roadmap/blob/master/design_drafts/hardware_access.md

### PR DESCRIPTION
**Pull request by Karsten1987**
Last part of #164 and closes #164 

Currently sits on top of #216 and addresses the comment in [here](https://github.com/ros-controls/ros2_control/pull/216#discussion_r518208708).

To start the node one can pass in a URDF as such:
```
ros2 run controller_manager ros2_control_node --ros-args -p update_rate:=10 -p "robot_description:=`cat install/test_robot_hardware/share/test_robot_hardware/test_robot_hardware.urdf`"
```

### CODE REVIEW

**Comment by Karsten1987** [comment]
Introduced a common URDF within the test_robot_hardware package, which reflects and exports all three types of hardware components.
The system can be loaded as such:
```
ros2 run controller_manager ros2_control_node --ros-args -p update_rate:=10 -p "robot_description:=`cat install/test_robot_hardware/share/test_robot_hardware/test_robot_hardware.urdf`"
```

Once the file is loaded, the system can be probed via a service call:
```
 ➭ ros2 control list_interfaces
command interfaces
	 base_joint/position
	 joint1/position
	 joint2/position
state interfaces
	 base_joint/position
	 ft_sensor/fx
	 ft_sensor/fy
	 ft_sensor/fz
	 ft_sensor/tx
	 ft_sensor/ty
	 ft_sensor/tz
	 joint1/position
	 joint2/position
```

**Comment by destogl** [comment]
@Karsten1987 Thanks for this!

I updated the demo example in ros2_controls/ros2_control_demos#37 to work with this.
Also, I changed the `forward_command_controller` to work.

To check the demo see [README](https://github.com/ros-controls/ros2_control_demos/blob/add_rrbot_system_position_joints/README.md#test-of-the-scenario-during-development-phase) instructions.

I had to create a cusom branches since I did small changes to this PR in #240 and a small chnage to `joint_state_controller` (see comment in ros-controls/ros2_control#109)

**Comment by Karsten1987** [comment]
I am finally putting this up for review. I've remodeled the controller manager and resource manager to load the hardware dynamically. That also revokes the dependency on the test_robot_hardware package, which could then eventually be reworked to store a common set of test materials, such as test controllers, test hardware components and their URDFs.

Please be aware that this PR will break the existing controller interfaces and all controllers within `ros2_controllers`.

File: controller_manager_msgs/srv/ListControllerInterfaces.srv [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (bmagyar) [inline controller_manager_msgs/srv/ListControllerInterfaces.srv:1]: I know it's a PITA but I'd strongly suggest renaming this service. We already have the name "controller interface" for a class, also it doesn't actually capture what's happening here, there's not even a controller at play in this context.

I suggest something along the lines of `ListJointInterfaces` but I wonder if this is strongly required at this stage... I think it may be useful to also denote whether these interfaces are claimed at the moment, perhaps by which controller, etc. I think that from the resource manager PR perspective, this addition is a bit of a feature creep that can lead to a lot more creeping as you can tell by my comment.

File: controller_interface/include/controller_interface/controller_interface.hpp [hunk 2]
```diff
@@ -44,12 +60,26 @@ class ControllerInterface : public rclcpp_lifecycle::node_interfaces::LifecycleN
...
+  CONTROLLER_INTERFACE_PUBLIC
+  virtual
+  InterfaceConfiguration command_interface_configuration() const = 0;
+
+  CONTROLLER_INTERFACE_PUBLIC
+  virtual
+  InterfaceConfiguration state_interface_configuration() const = 0;
+
+  CONTROLLER_INTERFACE_PUBLIC
+  void assign_interfaces(
```
Comment (bmagyar) [inline controller_interface/include/controller_interface/controller_interface.hpp:72]:
```suggestion
  void claim_interfaces(
```
We've used the word `claim` for this throughout ros_control

File: controller_interface/include/controller_interface/controller_interface.hpp [hunk 3]
```diff
@@ -34,6 +36,19 @@ enum class return_type : std::uint8_t
...
+{
+  ALL = 0,
+  INDIVIDUAL = 1,
+  NONE = 2,
+};
+
+struct InterfaceConfiguration
+{
+  configuration_type type;
+  std::vector<std::string> names = {};
```
Comment (bmagyar) [inline controller_interface/include/controller_interface/controller_interface.hpp:50]: names of what? joint names? interface names? combined joint + interface names?

File: controller_interface/include/controller_interface/controller_interface.hpp [hunk 4]
```diff
@@ -34,6 +36,19 @@ enum class return_type : std::uint8_t
   ERROR = 1,
 };
 
+enum class configuration_type : std::uint8_t
```
Comment (bmagyar) [inline controller_interface/include/controller_interface/controller_interface.hpp:40]: Um... This is super general and I have no clue what'd use this for. Please either document this, give those fields more descriptive names or let's think about alternative solutions.

File: controller_manager/include/controller_manager/controller_manager.hpp [hunk 5]
```diff
@@ -241,6 +270,8 @@ class ControllerManager : public rclcpp::Node
   /// mutex copied from ROS1 Control, protects service callbacks
   /// not needed if we're guaranteed that the callbacks don't come from multiple threads
   std::mutex services_lock_;
+  rclcpp::Service<controller_manager_msgs::srv::ListControllerInterfaces>::SharedPtr
+    list_controller_interfaces_service_;
```
Comment (destogl) [inline controller_manager/include/controller_manager/controller_manager.hpp:274]:
```suggestion
    list_resource_interfaces_service_;
```

File: controller_manager/src/controller_manager.cpp [hunk 6]
```diff
@@ -49,56 +50,104 @@ rclcpp::NodeOptions get_cm_node_options()
...
+  init_services();
+}
+
+ControllerManager::ControllerManager(
+  std::unique_ptr<hardware_interface::ResourceManager> resource_manager,
+  std::shared_ptr<rclcpp::Executor> executor,
+  const std::string & manager_node_name)
+: rclcpp::Node(manager_node_name, get_cm_node_options()),
+  resource_manager_(
+    std::shared_ptr<hardware_interface::ResourceManager>(resource_manager.release())),
```
Comment (destogl) [inline controller_manager/src/controller_manager.cpp:83]: This looks like a hack, since `resource_manager_` is a unique_ptr.

File: controller_interface/include/controller_interface/controller_interface.hpp (hunk 4 above)
Comment (destogl) [reply controller_interface/include/controller_interface/controller_interface.hpp:40]: Maybe something like this:
```suggestion
enum class interface_configuration_type : std::uint8_t
```

File: controller_interface/include/controller_interface/controller_interface.hpp (hunk 3 above)
Comment (destogl) [reply controller_interface/include/controller_interface/controller_interface.hpp:50]: It is always interface. There is no "joint" anymore here.
```suggestion
  std::vector<std::string> interface_names = {};
```

File: controller_interface/include/controller_interface/controller_interface.hpp (hunk 2 above)
Comment (destogl) [reply controller_interface/include/controller_interface/controller_interface.hpp:72]: @bmagyar I was at first with you on this. But later, when I saw CM, I am actually with @Karsten1987 on this.
This function is called from CM, which is actually assigning interfaces... This is semantically more correct. The logic in this concept is somewhat different, so that CM is assigning interfaces and not that controller claims them from hardware.

Claiming is happening [few lines before](https://github.com/ros-controls/ros2_control/pull/236/files#diff-51a3ffe225a0370e8b9026ea1cb8f0cf34fd5bf6547a239ce8418f445e22a094R637).

File: controller_manager/src/controller_manager.cpp [hunk 7]
```diff
@@ -618,6 +709,14 @@ void ControllerManager::start_controllers_asap()
 #endif
 }
 
+void ControllerManager::list_controller_interfaces_srv_cb(
```
Comment (destogl) [inline controller_manager/src/controller_manager.cpp:712]:
```suggestion
void ControllerManager::list_resource_interfaces_srv_cb(
```

Comment (destogl) [reply controller_manager/src/controller_manager.cpp:712]: We should also return the interfaces claimed status here. Very useful for introspection.

File: controller_manager/src/controller_manager.cpp [hunk 8]
```diff
@@ -811,9 +910,12 @@ std::vector<std::string> ControllerManager::get_controller_names()
   return names;
 }
 
-controller_interface::return_type
-ControllerManager::update()
+controller_interface::return_type ControllerManager::update(bool update_resources)
```
Comment (destogl) [inline controller_manager/src/controller_manager.cpp:913]: Why did you add this argument?

File: controller_manager/src/ros2_control_node.cpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    throw std::runtime_error("update_rate parameter not existing or empty");
+  }
+  RCLCPP_INFO(cm->get_logger(), "update rate is %d Hz", update_rate);
+
+  std::function<controller_interface::return_type()> timer_cb =
+    std::bind(&controller_manager::ControllerManager::update, cm.get(), true);
+  auto timer = cm->create_wall_timer(
+    std::chrono::milliseconds(1000 / update_rate),
+    timer_cb,
+    cm->deterministic_callback_group_);
```
Comment (destogl) [inline controller_manager/src/ros2_control_node.cpp:48]: As you commented in #216 I don't like this so much, but it seems to be the only way...

File: controller_manager_msgs/CMakeLists.txt [hunk 10]
```diff
@@ -18,6 +18,7 @@ set(msg_files
   msg/ControllerState.msg
 )
 set(srv_files
+  srv/ListControllerInterfaces.srv
```
Comment (destogl) [inline controller_manager_msgs/CMakeLists.txt:21]: Or maybe even better `ListHardwareInterfaces`
```suggestion
  srv/ListResourceInterfaces.srv
```

File: controller_manager_msgs/srv/ListControllerInterfaces.srv (hunk 1 above)
Comment (destogl) [reply controller_manager_msgs/srv/ListControllerInterfaces.srv:1]: I suggest: `ListHardwareInterfaces` or `ListResourceInterfaces`

I would leave it am make a new issue with these suggestions. It was very useful for testing of example implementation

File: hardware_interface/include/hardware_interface/resource_manager.hpp [hunk 11]
```diff
@@ -58,6 +61,18 @@ class ResourceManager
...
+  /**
+   * The resource manager can be post initialized with a given URDF.
+   * This is mainly used in conjunction with the default constructor
+   * in which the URDF might not be present at first initalization.
+   *
+   * \param urdf string containing the URDF.
+   * \param validate_interfaces boolean argument indicating whether the exported
+   * interfaces ought to be validated. Defaults to true.
+   */
+  void initialize_from_urdf(const std::string & urdf, bool validate_interfaces = true);
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/resource_manager.hpp:74]: Maybe better:
```suggestion
  void load_from_urdf(const std::string & urdf, bool validate_interfaces = true);
```

File: ros2controlcli/ros2controlcli/api/__init__.py [hunk 12]
```diff
@@ -56,6 +57,12 @@ def list_controllers(controller_manager_name):
                           ListControllers, request)
 
 
+def list_controller_interfaces(controller_manager_name):
```
Comment (destogl) [inline ros2controlcli/ros2controlcli/api/__init__.py:60]:
```suggestion
def list_hardware_interfaces(controller_manager_name):
```

File: ros2controlcli/ros2controlcli/verb/list_interfaces.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ros2cli.node.direct import add_arguments
+from ros2cli.verb import VerbExtension
+from ros2controlcli.api import add_controller_mgr_parsers, list_controller_interfaces
+
+
+class ListInterfacesVerb(VerbExtension):
```
Comment (destogl) [inline ros2controlcli/ros2controlcli/verb/list_interfaces.py:20]:
```suggestion
class ListHardwareInterfacesVerb(VerbExtension):
```

File: ros2controlcli/setup.py [hunk 14]
```diff
@@ -37,6 +37,7 @@
         ],
         'ros2controlcli.verb': [
             'list = ros2controlcli.verb.list:ListVerb',
+            'list_interfaces = ros2controlcli.verb.list_interfaces:ListInterfacesVerb',
```
Comment (destogl) [inline ros2controlcli/setup.py:40]:
```suggestion
            'list_hardware_interfaces = ros2controlcli.verb.list_interfaces:ListInterfacesVerb',
```

**Review by destogl — CHANGES_REQUESTED** [review]

File: ros2controlcli/setup.py (hunk 14 above)
Comment (Karsten1987) [reply ros2controlcli/setup.py]: done in ec56888

File: ros2controlcli/ros2controlcli/verb/list_interfaces.py (hunk 13 above)
Comment (Karsten1987) [reply ros2controlcli/ros2controlcli/verb/list_interfaces.py]: ec56888

File: ros2controlcli/ros2controlcli/api/__init__.py (hunk 12 above)
Comment (Karsten1987) [reply ros2controlcli/ros2controlcli/api/__init__.py]: ec56888

File: controller_manager_msgs/CMakeLists.txt (hunk 10 above)
Comment (Karsten1987) [reply controller_manager_msgs/CMakeLists.txt]: ec56888

File: controller_manager/src/controller_manager.cpp (hunk 7 above)
Comment (Karsten1987) [reply controller_manager/src/controller_manager.cpp]: ec56888

File: controller_manager/include/controller_manager/controller_manager.hpp (hunk 5 above)
Comment (Karsten1987) [reply controller_manager/include/controller_manager/controller_manager.hpp]: ec56888

File: controller_manager/include/controller_manager/controller_manager.hpp [hunk 15]
```diff
@@ -24,14 +24,15 @@
 
 #include "controller_manager/controller_spec.hpp"
 #include "controller_manager/visibility_control.h"
+#include "controller_manager_msgs/srv/list_controller_interfaces.hpp"
```
[the remark this answers is not in the document]
Comment (Karsten1987) [reply controller_manager/include/controller_manager/controller_manager.hpp]: ec56888

File: controller_manager_msgs/srv/ListControllerInterfaces.srv (hunk 1 above)
Comment (Karsten1987) [reply controller_manager_msgs/srv/ListControllerInterfaces.srv]: ec56888

renamed to `ListHardwareInterfaces`

File: controller_interface/src/controller_interface.cpp [hunk 16]
```diff
@@ -16,17 +16,19 @@
...
-  std::weak_ptr<hardware_interface::RobotHardware> robot_hardware,
-  const std::string & controller_name)
+ControllerInterface::init(const std::string & controller_name)
 {
-  robot_hardware_ = robot_hardware;
-  lifecycle_node_ = std::make_shared<rclcpp_lifecycle::LifecycleNode>(controller_name);
+  lifecycle_node_ = std::make_shared<rclcpp_lifecycle::LifecycleNode>(
+    controller_name,
+    rclcpp::NodeOptions().allow_undeclared_parameters(true).
+    automatically_declare_parameters_from_overrides(true));
```
[the remark this answers is not in the document]
Comment (Karsten1987) [reply controller_interface/src/controller_interface.cpp:31]: I would table this for #234

File: controller_manager/src/controller_manager.cpp (hunk 6 above)
Comment (Karsten1987) [reply controller_manager/src/controller_manager.cpp]: Good catch. Addressed in 4372131

File: hardware_interface/include/hardware_interface/resource_manager.hpp (hunk 11 above)
Comment (Karsten1987) [reply hardware_interface/include/hardware_interface/resource_manager.hpp]: done 50e58cf

File: controller_manager/src/controller_manager.cpp (hunk 8 above)
Comment (Karsten1987) [reply controller_manager/src/controller_manager.cpp:924]: well, that way a simple timer callback will take care of a complete cycle: `read, update, write`. However, some people might fancy more control over this and treat the controller manager solely responsible for updating the controllers. So the read/write can be done externally.

File: controller_interface/include/controller_interface/controller_interface.hpp (hunk 2 above)
Comment (Karsten1987) [reply controller_interface/include/controller_interface/controller_interface.hpp:71]: same here, claiming is not exposed in the controller interface anymore. The controller has no control over the controller manager neither the hardware.

File: controller_interface/include/controller_interface/controller_interface.hpp (hunk 3 above)
Comment (Karsten1987) [reply controller_interface/include/controller_interface/controller_interface.hpp:49]: Doesn't this sounds weird, "interface_configuration.interface_names"? That's kind of redundant in my opinion.
I like `interfae_configuration.names` somewhat more.

File: controller_interface/include/controller_interface/controller_interface.hpp (hunk 4 above)
Comment (Karsten1987) [reply controller_interface/include/controller_interface/controller_interface.hpp]: c5f65f9

File: controller_interface/src/controller_interface.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (destogl) [reply controller_interface/src/controller_interface.cpp:31]: I am not sure this is needed there anymore if we proceed as discussed on Wednesday

**Review by destogl — APPROVED** [review]
I am happy now :)
Thanks!

File: controller_manager/src/controller_manager.cpp (hunk 8 above)
Comment (bmagyar) [reply controller_manager/src/controller_manager.cpp:928]: I'd prefer if we didn't include this here, better keep things clean. We have the example node to take care of this.

**Comment by bmagyar** [comment]
What are the plans with the packages the build is broken on?

**Comment by destogl** [comment]
I see only an issue with joint_limits_interface. transmission_interface will change anyway and it should be kind of independent from all this.

Are there any other issues you see?

**Comment by Karsten1987** [comment]
I think that the joint limit interface can safely be ignored for now. I believe that we actually don't need a dedicated package for this. The functionality can easily be incorporated within either the hardware_interface or into a default component package. Interface seems to be wrong in this case then.

File: controller_interface/include/controller_interface/controller_interface.hpp [hunk 17]
```diff
@@ -44,12 +66,26 @@ class ControllerInterface : public rclcpp_lifecycle::node_interfaces::LifecycleN
...
+  CONTROLLER_INTERFACE_PUBLIC
+  void release_interfaces();
+
   CONTROLLER_INTERFACE_PUBLIC
   virtual
   return_type
-  init(
-    std::weak_ptr<hardware_interface::RobotHardware> robot_hardware,
-    const std::string & controller_name);
+  init(const std::string & controller_name);
```
Comment (destogl) [inline controller_interface/include/controller_interface/controller_interface.hpp:88]: Reviewing the current controller portings in ros-control/ros2_controllers repository I realized, that this name argument is obsolete. Since controllers are `LifecycleNode`s they have access to their own name through the node name.

```suggestion
  init();
```

Comment (bmagyar) [reply controller_interface/include/controller_interface/controller_interface.hpp:88]: #246 says "Hold my beer"

**Comment by guru-florida** [comment]
I updated the code for joint limit interface in PR #271. Discussion about moving these files into hardware_interface is in there.

> I think that the joint limit interface can safely be ignored for now. I believe that we actually don't need a dedicated package for this. The functionality can easily be incorporated within either the hardware_interface or into a default component package. Interface seems to be wrong in this case then.
